aten/src/ATen/quantized/Quantizer.h

```diff
@@ -0,0 +1,251 @@
+ * of weights since per-input channel of weight quantization or per-channel
+ * quantization for activations can't be efficiently supported in most of
+ * processors since it requires each multiplication result within a single
+ * dot-product to have a different scale.
+ */
+struct CAFFE2_API PerChannelSymmetricQuantizer : public SymmetricQuantizer {
+  explicit PerChannelSymmetricQuantizer(
+      const std::vector<float> scales,
+      const std::vector<int64_t> axis)
+      : SymmetricQuantizer(kPerChannelSymmetric), scales_(scales), axis_(axis) {
+    AT_CHECK(
+        axis_.size() == 1,
+        "Per channel symmetric quantization in multiple axis is not supported yet.");
+  }
+
+  std::vector<float> scales() const {
+    return scales_;
+  }
+
+  std::vector<int64_t> axis() const {
+    return axis_;
+  }
+
+ private:
+  const std::vector<float> scales_;
+  const std::vector<int64_t> axis_;
+};
+
+/**
+ * PerTensorAffineQuantizer stores a scale and a zero_point, which is used for
+ * all the values in the Tensor.
+ */
+struct CAFFE2_API PerTensorAffineQuantizer : public AffineQuantizer {
+  explicit PerTensorAffineQuantizer(float scale, uint8_t zero_point)
+      : AffineQuantizer(kPerTensorAffine),
+        scale_(scale),
+        zero_point_(zero_point) {}
+
+  virtual QTensor quantize(RealTensor tensor);
+  virtual RealTensor dequantize(QTensor tensor);
+
+  float scale() const {
+    return scale_;
+  }
+
+  uint8_t zero_point() const {
+    return zero_point_;
+  }
+
+ private:
+  const float scale_;
+  const uint8_t zero_point_;
+};
+
+/**
+ * PerChannelAffineQuantizer is the same as PerTensorAffineQuantizer
+ * except that we have an independent scale and zero_point parameter
+ * for each channel.
+ */
+struct CAFFE2_API PerChannelAffineQuantizer : public AffineQuantizer {
+  explicit PerChannelAffineQuantizer(
+      const std::vector<float> scales,
+      const std::vector<uint8_t> zero_points,
+      const std::vector<int64_t> axis)
+      : AffineQuantizer(kPerChannelAffine),
+        scales_(scales),
+        zero_points_(zero_points),
+        axis_(axis) {
+    AT_CHECK(
+        axis_.size() == 1,
+        "Per channel affine quantization in multiple axis is not supported yet.");
+  }
+
+  std::vector<float> scales() const {
+    return scales_;
+  }
+
+  std::vector<uint8_t> zero_points() const {
+    return zero_points_;
+  }
+
+  std::vector<int64_t> axis() const {
+    return axis_;
+  }
+
+ private:
+  const std::vector<float> scales_;
+  const std::vector<uint8_t> zero_points_;
+  const std::vector<int64_t> axis_;
+};
+
+// This is an internal utility function for getting at the QTensorImpl,
+// You should only use this for writing low level
+// setters/getters for QTensorImpl fields; otherwise, you should use
+// the low level setters/getters that were implemented using this.
+// This may be called repeatedly, so make sure it's pretty cheap.
+CAFFE2_API QTensorImpl* get_qtensorimpl(const QTensor& self);
```

Comment:
Would it make sense to make this a private method of `QTensor` (and adding all users to `friend`s)?

Reply:
We don't have a separate QTensor class, it's the same as Tensor..

Reply:
Ah, please ignore this then :)

Reply:
I'm not sure if it's a good idea to make QTensorImpl as a friend class of Tensor, which might contain a intrusive_ptr to QTensorImpl